sequencing-server/src/utils/hasura.ts

```diff
@@ -0,0 +1,314 @@
+ * @param authHeader If the user sends a POST request directly to the endpoint with a bearer token this will be populated.
+ * @returns The username of the user sending the request.
+ */
+export function getUsername(
+  authSessionVariables: AuthSessionVariables | undefined,
+  authHeader: string | undefined,
+): string {
+  if (authSessionVariables !== undefined) {
+    return authSessionVariables['x-hasura-user-id'];
+  } else if (authHeader !== undefined) {
+    return (jwtDecode(authHeader) as Jwt).username;
+  }
+
+  throw new Error('Could not determine the user sending the request');
+}
+
+/**
+ * This function checks to see if request coming from the Hasura action is being executed
+ * by a user with enough permissions. To check this we need to call a GQL query with the
+ * action key and the user session. If we ever merge the databases this should be changed
+ * to call the db function that does the permission checking directly.
+ *
+ * Some endpoints don't need to be checked which is why we have the mappings of endpoints
+ * to actions to check.
+ *
+ * @param url The endpoint that we're checking the permissions for.
+ * @param graphqlClient The GQL Client we're using to make the permission check.
+ * @returns True if the user is able to call the given action, false otherwise.
+ */
+export async function canUserPerformAction(
+  url: string,
+  graphqlClient: GraphQLClient,
+  hasuraSession: AuthSessionVariables,
+  body: any,
+): Promise<boolean> {
+  const role = hasuraSession['x-hasura-role'];
+  const username = hasuraSession['x-hasura-user-id'];
+
+  // The aerie_admin role always has NO_CHECK permissions on all actions.
+  if (role === 'aerie_admin') {
+    return true;
+  }
+
+  const permissionCheckQuery = await graphqlClient.request(
+    gql`
+      query checkUserPermissions($role: user_roles_enum!, $key: String!) {
+        user_role_permission_by_pk(role: $role) {
+          action_permissions(path: $key)
+        }
+      }
+    `,
+    {
+      role,
+      key: ENDPOINTS_TO_ACTION_KEY[url],
+    },
+  );
+
+  const permission = permissionCheckQuery.user_role_permission_by_pk.action_permissions as HasuraPermissions | null;
+
+  if (permission === null) {
+    return false;
+  }
+
+  const simulationDatasetId = body.input.simulationDatasetId as number;
+  let missionModelId = body.input.missionModelId as number;
+  let planId = body.input.planId as number;
+
+  // If we have a missionModelId and we need to check a plan permission, get the planId.
+  if (missionModelId !== null && permission in HASURA_PLAN_PERMISSIONS) {
+    planId = await getPlanId(graphqlClient, missionModelId);
+  }
+
+  // If we have a simulationDatasetId and we need to check a plan permission, get the planId.
+  if (simulationDatasetId !== null && permission in HASURA_PLAN_PERMISSIONS) {
+    planId = await getPlanId(graphqlClient, undefined, simulationDatasetId);
+  }
```

Comment:
Why only do this if `permission in HASURA_PLAN_PERMISSIONS`? The code would be simpler if you were always working with `planId`s.

Reply:
I'm trying to avoid extraneous calls, if an action gets called with a `simulationDatasetId` I only try and fetch the plan id for that simulation dataset if the user is actually checking one of the plan permissions.
`NO_CHECK` won't do another call
`MISSION_MODEL_OWNER` will get the mission model.
The rest will fallback to using the `planId`.

I missed adding `OWNER` to `HASURA_PLAN_PERMISSIONS` though so I'll fix that.

Reply:
Regarding the `MISSION_MODEL_OWNER` case: I have mixed feelings. While I agree that saving a query is a good thing, I think it does so at the detriment to the readability of the function `getMissionModelId`, which is really two disjoint functions stitched together via optional arguments.